### PR TITLE
[EN] #4224 Check Chunk Data Pack request comes from staked node

### PR DESCRIPTION
**Pull request by m4ksio**

### CODE REVIEW

**Review by ramtinms — APPROVED** [review]

File: engine/execution/provider/engine.go [hunk 1]
```diff
@@ -168,7 +158,26 @@ func (e *Engine) onChunkDataRequest(
...
+		return engine.NewInvalidInputErrorf("invalid origin id (%s): %w", origin, err)
+	}
+
+	// only verifier nodes are allowed to request chunk data packs
+	if origin.Role != flow.RoleVerification {
+		return engine.NewInvalidInputErrorf("invalid role for receiving collection: %s", origin.Role)
+	}
+
+	if origin.Stake == 0 {
+		return engine.NewInvalidInputErrorf("node %s is not staked for the epoch corresponding to the requested chunk data pack", origin.NodeID)
```
Comment (yhassanzadeh13) [inline engine/execution/provider/engine.go:180]: This part can be encapsulated in a method [like this](https://github.com/onflow/flow-go/pull/427/files#diff-e2a1bac627623a345dc4e9caf84fa096cdecddbe535730904d7438062ffb3820R246-R259). The advantage is that the returned error can be logged at the invoker (i.e., `onChunkDataRequest`) accompanied with more details about the chunk data pack request.

File: engine/execution/provider/engine_test.go [hunk 2]
```diff
@@ -15,6 +15,7 @@ import (
 	"github.com/onflow/flow-go/model/messages"
 	"github.com/onflow/flow-go/module/metrics"
 	"github.com/onflow/flow-go/network/mocknetwork"
+	protocol2 "github.com/onflow/flow-go/state/protocol"
 	protocol "github.com/onflow/flow-go/state/protocol/mock"
```
Comment (yhassanzadeh13) [inline engine/execution/provider/engine_test.go:19]: I would refactor like this:
```
"github.com/onflow/flow-go/state/protocol"
mockprotocol "github.com/onflow/flow-go/state/protocol/mock"
```

File: storage/headers.go [hunk 3]
```diff
@@ -24,4 +24,10 @@ type Headers interface {
 	// be unfinalized; if there is more than one, at least one of them has to
 	// be unfinalized.
 	ByParentID(parentID flow.Identifier) ([]*flow.Header, error)
+
+	// Indexes block ID by chunk ID
+	IndexByChunkID(headerID, chunkID flow.Identifier) error
+
+	// Finds the ID of the block containing given collection
+	IDByChunkID(chunkID flow.Identifier) (flow.Identifier, error)
```
Comment (yhassanzadeh13) [inline storage/headers.go:32]: I would name this method as `BlockID(chunkID flow.Identifier) ....`.

File: storage/badger/headers.go [hunk 4]
```diff
@@ -55,6 +62,15 @@ func NewHeaders(collector module.CacheMetrics, db *badger.DB) *Headers {
 		}
 	}
 
+	retrieveChunkID := func(key interface{}) func(tx *badger.Txn) (interface{}, error) {
+		chunkID := key.(flow.Identifier)
+		var blockID flow.Identifier
```
Comment (yhassanzadeh13) [inline storage/badger/headers.go:67]: do we need this closure? 🤔

File: engine/execution/provider/engine_test.go [hunk 5]
```diff
@@ -48,6 +57,78 @@ func TestProviderEngine_onChunkDataRequest(t *testing.T) {
...
+			Return(&flow.ChunkDataPack{CollectionID: collectionID, ChunkID: chunkID}, nil)
+		execState.
+			On("GetBlockIDByChunkID", chunkID).
+			Return(blockID, nil)
+
+		req := &messages.ChunkDataRequest{
+			ChunkID: chunkID,
+			Nonce:   rand.Uint64(),
+		}
+		// submit using origin ID with zero stake
```
Comment (yhassanzadeh13) [inline engine/execution/provider/engine_test.go:123]:
```suggestion
		// submit using non-existing origin ID
```

File: engine/execution/state/state.go [hunk 6]
```diff
@@ -460,6 +470,10 @@ func (s *state) GetCollection(identifier flow.Identifier) (*flow.Collection, err
 	return s.collections.ByID(identifier)
 }
 
+func (s *state) GetBlockIDByChunkID(collectionID flow.Identifier) (flow.Identifier, error) {
```
Comment (yhassanzadeh13) [inline engine/execution/state/state.go:473]: shouldn't `collectionID` be `chunkID`?

File: storage/headers.go (hunk 3 above)
Comment (yhassanzadeh13) [reply storage/headers.go:32]: Also, godoc may need refactoring (collection?)

**Review by yhassanzadeh13 — APPROVED** [review]

File: engine/execution/provider/engine_test.go (hunk 2 above)
Comment (m4ksio) [reply engine/execution/provider/engine_test.go:19]: Great spot, fixed

File: storage/headers.go (hunk 3 above)
Comment (m4ksio) [reply storage/headers.go:32]: But we return only Identifier not whole object, name like `BlockID` would be misleading.
Fixed the parameter name though

File: storage/badger/headers.go (hunk 4 above)
Comment (m4ksio) [reply storage/badger/headers.go:67]: Its used in chunk cache

File: engine/execution/state/state.go (hunk 6 above)
Comment (m4ksio) [reply engine/execution/state/state.go:473]: It should!

**Pull request merged**